opm/material/fluidsystems/blackoilpvt/DryHumidGasPvt.hpp

```diff
@@ -0,0 +1,666 @@
+
+                invGasB.appendXPos(pg);
+                gasMu.appendXPos(pg);
+
+                invSatGasBArray.push_back(1.0/B);
+                invSatGasBMuArray.push_back(1.0/(mu*B));
+
+                assert(invGasB.numX() == outerIdx + 1);
+                assert(gasMu.numX() == outerIdx + 1);
+
+                const auto& underSaturatedTable = pvtgwTable.getUnderSaturatedTable(outerIdx);
+                size_t numRows = underSaturatedTable.numRows();
+                for (size_t innerIdx = 0; innerIdx < numRows; ++ innerIdx) {
+                    Scalar Rw = underSaturatedTable.get("RW" , innerIdx);
+                    Scalar Bg = underSaturatedTable.get("BG" , innerIdx);
+                    Scalar mug = underSaturatedTable.get("MUG" , innerIdx);
+
+                    invGasB.appendSamplePoint(outerIdx, Rw, 1.0/Bg);
+                    gasMu.appendSamplePoint(outerIdx, Rw, mug);
+                }
+            }
+
+            {
+                std::vector<double> tmpPressure =  saturatedTable.getColumn("PG").vectorCopy( );
+
+                invSatGasB.setXYContainers(tmpPressure, invSatGasBArray);
+                invSatGasBMu.setXYContainers(tmpPressure, invSatGasBMuArray);
+            }
+
+            // make sure to have at least two sample points per gas pressure value
+            for (unsigned xIdx = 0; xIdx < invGasB.numX(); ++xIdx) {
+               // a single sample point is definitely needed
+                assert(invGasB.numY(xIdx) > 0);
+
+                // everything is fine if the current table has two or more sampling points
+                // for a given mole fraction
+                if (invGasB.numY(xIdx) > 1)
+                    continue;
+
+                // find the master table which will be used as a template to extend the
+                // current line. We define master table as the first table which has values
+                // for undersaturated gas...
+                size_t masterTableIdx = xIdx + 1;
+                for (; masterTableIdx < saturatedTable.numRows(); ++masterTableIdx)
+                {
+                    if (pvtgwTable.getUnderSaturatedTable(masterTableIdx).numRows() > 1)
+                        break;
+                }
+
+                if (masterTableIdx >= saturatedTable.numRows())
+                    throw std::runtime_error("PVTGW tables are invalid: The last table must exhibit at least one "
+                              "entry for undersaturated gas!");
+
+
+                // extend the current table using the master table.
+                extendPvtgwTable_(regionIdx,
+                                 xIdx,
+                                 pvtgwTable.getUnderSaturatedTable(xIdx),
+                                 pvtgwTable.getUnderSaturatedTable(masterTableIdx));
+            }
+        }
+
+        vapPar1_ = 0.0;
+        
+        initEnd();
+    }
+
+private:
+    void extendPvtgwTable_(unsigned regionIdx,
+                          unsigned xIdx,
```

Comment:
The method `extendPvtgwTable_` (similar methods exists for other PVT tables) comes in when the `PVTGW` table has no undersaturated table for a provided pressure value (say `P`). The method has a saturated table  (`masterTable`) as input. This is the first undersaturated table of the `PVTGW` table. An undersaturated table  will be created for pressure `P` by extrapolating  the saturated values at  `P` and the extrapolated values are such that the relative changes for Bg and viscosity, expressed in difference/average, are the same as for the` masterTable`. The relative change is referred to by "compressibility"" and "viscosibility" in the comments which is maybe a bit misleading because the difference and average is taken over `RW` values, not over pressures.

Reply:
Thanks  for the explanations.